sdk/src/main/java/com/qonversion/android/sdk/QProductCenterManager.kt

```diff
@@ -513,33 +556,118 @@ class QProductCenterManager internal constructor(
         val callbacks = permissionsCallbacks.toList()
         permissionsCallbacks.clear()
 
+        retryLaunchForPermissions(
+            { permissions -> handleSuccessPermissions(callbacks, permissions) },
+            { error -> handleFailurePermissions(callbacks, error) }
+        )
+    }
+
+    private fun retryLaunchForProducts(onCompleted: () -> Unit) {
+        launchResult?.let {
+            handleLoadStateForProducts(onCompleted)
+        } ?: retryLaunch(
+            onSuccess = {
+                handleLoadStateForProducts(onCompleted)
+            },
+            onError = {
+                handleLoadStateForProducts(onCompleted)
+            })
+    }
+
+    private fun retryLaunch(
+        onSuccess: (QLaunchResult) -> Unit,
+        onError: (QonversionError) -> Unit
+    ) {
+        launch(object : QonversionLaunchCallback {
+            override fun onSuccess(launchResult: QLaunchResult) = onSuccess(launchResult)
+            override fun onError(error: QonversionError) = onError(error)
+        })
+    }
+
+    private fun handleLoadStateForProducts(onCompleted: () -> Unit) {
+        when (loadProductsState) {
+            Loaded -> onCompleted()
+            Failed -> loadStoreProductsIfPossible()
+            else -> Unit
+        }
+    }
+
+    private fun handleSuccessProducts(
```

Comment:
Looks like this function used only once. Maybe you should remove and use this code block directly